src/tracing/logger/index.ts

```diff
@@ -26,25 +26,64 @@ export function init(opts: LoggerOptions): Logger {
 
 
 export namespace Tracing {
-    type DecoratorFn = (target: unknown, propertyKey: string, descriptor: PropertyDescriptor) => void;
+    type DecoratorFn<T = any> = (target: any, propertyKey: string, descriptor: TypedPropertyDescriptor<T>) => void;
 
     /** Retrieve the current span if any */
     export function currentSpan(): Span | undefined {
         return logger?.currentSpan();
     }
 
-    export function traceWithTests<T>(name: string, tests: Test<T>[], callback: () => T): T {
-        if (!logger) {
-            return callback();
+    /** 
+     * Trace a function call and run tests on the result
+     * 
+     * @template T is the type of the result (and the input of the tests)
+     * @param tests The tests to run on the result
+     * @param callback The function to trace
+     * @returns The result of the function
+     */
+    export function traceWithTests<T extends (...args: any[]) => any, U extends ReturnType<T>>(name: string, tests: Test<U>[]): DecoratorFn<T>;
+    export function traceWithTests<T extends (...args: any[]) => any, U extends ReturnType<T>>(tests: Test<U>[]): DecoratorFn<T>;
+    export function traceWithTests<T extends (...args: any[]) => any, U extends ReturnType<T>>(name: string, tests: Test<U>[], callback: T): U;
+    export function traceWithTests<T extends (...args: any[]) => any, U extends ReturnType<T>>(tests: Test<U>[], callback: T): U;
+    export function traceWithTests<T extends (...args: any[]) => any, U extends ReturnType<T>>(nameOrTests: string | Test<U>[], testsOrCallback?: Test<U>[] | T, callback?: T): U | DecoratorFn<T> {
```

Comment:
the tests passed into the decorator will be type-checked to match the output of the decorated function 😎